AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -3,43 +3,129 @@
+                    completion(tokens, error);
+                }
+            }];
+    }
+    else
+    {
+        NSError *error;
+        completion([[MSTokensResponse alloc] initWithTokens:@[cachedToken]], error);
+    }
+}
+
++ (void)saveToken:(MSTokenResult *)tokenResult {
+    NSString *tokenString = [tokenResult serializeToString];
+    BOOL success = [MSKeychainUtil storeString:tokenString forKey:[MSTokenExchange tokenKeyNameForPartition:tokenResult.partition]];
+    if (success) {
+        MSLogDebug([MSDataStore logTag], @"Saved token in keychain for the partitionKey : %@.", tokenResult.partition);
+    } else {
+        MSLogWarning([MSDataStore logTag], @"Failed to save the token in keychain for the partitionKey : %@.", tokenResult.partition);
+    }
+}
 
-  // Payload.
-  NSError *jsonError;
-  NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : partitions} options:0 error:&jsonError];
-
-  // Http call.
-  [httpClient sendAsync:payloadData
-      completionHandler:^(NSString *callId, NSHTTPURLResponse *response, NSData *data, NSError *error) {
-        MSLogVerbose([MSDataStore logTag], @"Get token callback, request Id %@ with status code: %lu", callId,
-                     (unsigned long)response.statusCode);
-
-        // If comletion is provided.
-        if (completion) {
-
-          // Read tokens.
-          NSError *tokenResponsejsonError;
-          NSDictionary *jsonDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&tokenResponsejsonError];
-          if (tokenResponsejsonError) {
-            MSLogError([MSDataStore logTag], @"Can't deserialize tokens with error: %@", [tokenResponsejsonError description]);
-            completion([[MSTokensResponse alloc] initWithTokens:nil], error);
-          }
-
-          // Create token result object.
-          MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
-          completion(tokens, error);
++ (MSTokenResult *)retrieveCachedToken:(NSString *)partitionName {
+    NSString *tokenString = [MSKeychainUtil stringForKey:[MSTokenExchange tokenKeyNameForPartition:partitionName]];
+    if(tokenString != nil){
+        MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithString:tokenString];
+        
+        if(tokenResult != nil){
+            
+            NSDate *currentUTCDate = [NSDate date];
+            
+            NSDateFormatter* utcFormatter = [[NSDateFormatter alloc] init];
+            [utcFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
+            NSDate *tokenExpireDate = [utcFormatter dateFromString:tokenResult.expiresOn];
+            
+            if([currentUTCDate laterDate:tokenExpireDate] == currentUTCDate){
+                [MSTokenExchange removeCachedToken:partitionName];
+                
+                 MSLogWarning([MSDataStore logTag], @"The token in the cache has expired for the partitionKey : %@.", partitionName);
```

Comment:
I don't know what is the correct log level for this case but seems it isn't right.
What's the SDK behavior if it failed to retrieve cached token here?

Reply:
If the token is not cached we need to make a call to the token exchange service to get a new token. The assumption is either the token has expired and deleted from the cache or it was never cached in the first place hence the http call to the service